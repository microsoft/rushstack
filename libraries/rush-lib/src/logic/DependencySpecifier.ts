// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import npmPackageArg from 'npm-package-arg';
import { InternalError } from '@rushstack/node-core-library';

/**
 * match workspace protocol in dependencies value declaration in `package.json`
 * example:
 * `"workspace:*"`
 * `"workspace:alias@1.2.3"`
 */
const WORKSPACE_PREFIX_REGEX: RegExp = /^workspace:((?<alias>[^._/][^@]*)@)?(?<version>.*)$/;

/**
 * resolve workspace protocol(from `@pnpm/workspace.spec-parser`).
 * used by pnpm. see [pkgs-graph](https://github.com/pnpm/pnpm/blob/27c33f0319f86c45c1645d064cd9c28aada80780/workspace/pkgs-graph/src/index.ts#L49)
 */
class WorkspaceSpec {
  public readonly alias?: string;
  public readonly version: string;
  public readonly versionSpecifier: string;

  public constructor(version: string, alias?: string) {
    this.version = version;
    this.alias = alias;
    this.versionSpecifier = alias ? `${alias}@${version}` : version;
  }

  public static tryParse(pref: string): WorkspaceSpec | undefined {
    const parts: RegExpExecArray | null = WORKSPACE_PREFIX_REGEX.exec(pref);
    if (parts?.groups) {
      return new WorkspaceSpec(parts.groups.version, parts.groups.alias);
    }
  }

  public toString(): `workspace:${string}` {
    return `workspace:${this.versionSpecifier}`;
  }
}

/**
 * The parsed format of a provided version specifier.
 */
export enum DependencySpecifierType {
  /**
   * A git repository
   */
  Git = 'Git',

  /**
   * A tagged version, e.g. "example@latest"
   */
  Tag = 'Tag',

  /**
   * A specific version number, e.g. "example@1.2.3"
   */
  Version = 'Version',

  /**
   * A version range, e.g. "example@2.x"
   */
  Range = 'Range',

  /**
   * A local .tar.gz, .tar or .tgz file
   */
  File = 'File',

  /**
   * A local directory
   */
  Directory = 'Directory',

  /**
   * An HTTP url to a .tar.gz, .tar or .tgz file
   */
  Remote = 'Remote',

  /**
   * A package alias, e.g. "npm:other-package@^1.2.3"
   */
  Alias = 'Alias',

  /**
   * A package specified using workspace protocol, e.g. "workspace:^1.2.3"
   */
  Workspace = 'Workspace'
}

/**
 * An NPM "version specifier" is a string that can appear as a package.json "dependencies" value.
 * Example version specifiers: `^1.2.3`, `file:./blah.tgz`, `npm:other-package@~1.2.3`, and so forth.
 * A "dependency specifier" is the version specifier information, combined with the dependency package name.
 */
export class DependencySpecifier {
  /**
   * The dependency package name, i.e. the key from a "dependencies" key/value table.
   */
  public readonly packageName: string;

  /**
   * The dependency version specifier, i.e. the value from a "dependencies" key/value table.
   * Example values: `^1.2.3`, `file:./blah.tgz`, `npm:other-package@~1.2.3`
   */
  public readonly versionSpecifier: string;

  /**
   * The type of the `versionSpecifier`.
   */
  public readonly specifierType: DependencySpecifierType;

  /**
   * If `specifierType` is `alias`, then this is the parsed target dependency.
   * For example, if version specifier i `"npm:other-package@^1.2.3"` then this is the parsed object for
   * `other-package@^1.2.3`.
   */
  public readonly aliasTarget: DependencySpecifier | undefined;

  public constructor(packageName: string, versionSpecifier: string) {
    this.packageName = packageName;
    this.versionSpecifier = versionSpecifier;

    // Workspace ranges are a feature from PNPM and Yarn. Set the version specifier
    // to the trimmed version range.
    const workspaceSpecResult: WorkspaceSpec | undefined = WorkspaceSpec.tryParse(versionSpecifier);
    if (workspaceSpecResult) {
      this.specifierType = DependencySpecifierType.Workspace;
      this.versionSpecifier = workspaceSpecResult.versionSpecifier;

      if (workspaceSpecResult.alias) {
        // "workspace:some-package@^1.2.3" should be resolved as alias
        this.aliasTarget = new DependencySpecifier(workspaceSpecResult.alias, workspaceSpecResult.version);
      } else {
        this.aliasTarget = undefined;
      }

      return;
    }

    const result: npmPackageArg.Result = npmPackageArg.resolve(packageName, versionSpecifier);
    this.specifierType = DependencySpecifier.getDependencySpecifierType(result.type);

    if (this.specifierType === DependencySpecifierType.Alias) {
      const aliasResult: npmPackageArg.AliasResult = result as npmPackageArg.AliasResult;
      if (!aliasResult.subSpec || !aliasResult.subSpec.name) {
        throw new InternalError('Unexpected result from npm-package-arg');
      }
      this.aliasTarget = new DependencySpecifier(aliasResult.subSpec.name, aliasResult.subSpec.rawSpec);
    } else {
      this.aliasTarget = undefined;
    }
  }

  public static getDependencySpecifierType(specifierType: string): DependencySpecifierType {
    switch (specifierType) {
      case 'git':
        return DependencySpecifierType.Git;
      case 'tag':
        return DependencySpecifierType.Tag;
      case 'version':
        return DependencySpecifierType.Version;
      case 'range':
        return DependencySpecifierType.Range;
      case 'file':
        return DependencySpecifierType.File;
      case 'directory':
        return DependencySpecifierType.Directory;
      case 'remote':
        return DependencySpecifierType.Remote;
      case 'alias':
        return DependencySpecifierType.Alias;
      default:
        throw new InternalError(`Unexpected npm-package-arg result type "${specifierType}"`);
    }
  }
}
