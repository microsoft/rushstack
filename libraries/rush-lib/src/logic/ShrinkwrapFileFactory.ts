// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import type { PackageManagerName } from '../api/packageManager/PackageManager';
import type { BaseShrinkwrapFile } from './base/BaseShrinkwrapFile';
import { NpmShrinkwrapFile } from './npm/NpmShrinkwrapFile';
import { PnpmShrinkwrapFile } from './pnpm/PnpmShrinkwrapFile';
import { YarnShrinkwrapFile } from './yarn/YarnShrinkwrapFile';

export class ShrinkwrapFileFactory {
  public static getShrinkwrapFile(
    packageManager: PackageManagerName,
    shrinkwrapFilename: string
  ): BaseShrinkwrapFile | undefined {
    switch (packageManager) {
      case 'npm':
        return NpmShrinkwrapFile.loadFromFile(shrinkwrapFilename);
      case 'pnpm':
        return PnpmShrinkwrapFile.loadFromFile(shrinkwrapFilename);
      case 'yarn':
        return YarnShrinkwrapFile.loadFromFile(shrinkwrapFilename);
      default:
        throw new Error(`Invalid package manager: ${packageManager}`);
    }
  }

  public static parseShrinkwrapFile(
    packageManager: PackageManagerName,
    shrinkwrapContent: string
  ): BaseShrinkwrapFile | undefined {
    switch (packageManager) {
      case 'npm':
        return NpmShrinkwrapFile.loadFromString(shrinkwrapContent);
      case 'pnpm':
        return PnpmShrinkwrapFile.loadFromString(shrinkwrapContent);
      case 'yarn':
        return YarnShrinkwrapFile.loadFromString(shrinkwrapContent);
      default:
        throw new Error(`Invalid package manager: ${packageManager}`);
    }
  }
}
