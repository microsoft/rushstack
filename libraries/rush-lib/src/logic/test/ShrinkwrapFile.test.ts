// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import * as path from 'path';
import { JsonFile } from '@rushstack/node-core-library';

import type { BaseShrinkwrapFile } from '../base/BaseShrinkwrapFile';
import { ShrinkwrapFileFactory } from '../ShrinkwrapFileFactory';
import {
  parsePnpmDependencyKey,
  PnpmShrinkwrapFile,
  ShrinkwrapFileMajorVersion
} from '../pnpm/PnpmShrinkwrapFile';
import { DependencySpecifier } from '../DependencySpecifier';
import { NpmShrinkwrapFile } from '../npm/NpmShrinkwrapFile';
import type { RushConfigurationProject } from '../../api/RushConfigurationProject';

describe(NpmShrinkwrapFile.name, () => {
  const filename: string = `${__dirname}/shrinkwrapFile/npm-shrinkwrap.json`;
  const shrinkwrapFile: BaseShrinkwrapFile = ShrinkwrapFileFactory.getShrinkwrapFile('npm', filename)!;

  it('verifies root-level dependency', () => {
    expect(shrinkwrapFile.hasCompatibleTopLevelDependency(new DependencySpecifier('q', '~1.5.0'))).toEqual(
      true
    );
  });

  it('verifies temp project dependencies', () => {
    // Found locally
    expect(
      shrinkwrapFile.tryEnsureCompatibleDependency(
        new DependencySpecifier('jquery', '>=2.2.4 <3.0.0'),
        '@rush-temp/project2'
      )
    ).toEqual(true);
    // Found at root
    expect(
      shrinkwrapFile.tryEnsureCompatibleDependency(
        new DependencySpecifier('q', '~1.5.0'),
        '@rush-temp/project2'
      )
    ).toEqual(true);
  });

  it('extracts temp projects successfully', () => {
    const tempProjectNames: ReadonlyArray<string> = shrinkwrapFile.getTempProjectNames();

    expect(tempProjectNames).toEqual(['@rush-temp/project1', '@rush-temp/project2']);
  });
});

describe(PnpmShrinkwrapFile.name, () => {
  describe('non-workspace', () => {
    function validateNonWorkspaceLockfile(shrinkwrapFile: BaseShrinkwrapFile): void {
      it('verifies root-level dependency', () => {
        expect(
          shrinkwrapFile.hasCompatibleTopLevelDependency(new DependencySpecifier('q', '~1.5.0'))
        ).toEqual(false);
      });

      it('verifies temp project dependencies', () => {
        expect(
          shrinkwrapFile.tryEnsureCompatibleDependency(
            new DependencySpecifier('jquery', '>=1.0.0 <2.0.0'),
            '@rush-temp/project1'
          )
        ).toEqual(true);
        expect(
          shrinkwrapFile.tryEnsureCompatibleDependency(
            new DependencySpecifier('q', '~1.5.0'),
            '@rush-temp/project2'
          )
        ).toEqual(true);
        expect(
          shrinkwrapFile.tryEnsureCompatibleDependency(
            new DependencySpecifier('pad-left', '^2.0.0'),
            '@rush-temp/project1'
          )
        ).toEqual(false);

        if (
          shrinkwrapFile instanceof PnpmShrinkwrapFile &&
          shrinkwrapFile.shrinkwrapFileMajorVersion >= ShrinkwrapFileMajorVersion.V9
        ) {
          expect(
            shrinkwrapFile.tryEnsureCompatibleDependency(
              new DependencySpecifier(
                '@scope/testDep',
                'https://github.com/jonschlinkert/pad-left/tarball/2.1.0'
              ),
              '@rush-temp/project3'
            )
          ).toEqual(true);
        } else {
          expect(
            shrinkwrapFile.tryEnsureCompatibleDependency(
              new DependencySpecifier('@scope/testDep', '>=2.0.0 <3.0.0'),
              '@rush-temp/project3'
            )
          ).toEqual(true);
        }
      });

      it('extracts temp projects successfully', () => {
        const tempProjectNames: ReadonlyArray<string> = shrinkwrapFile.getTempProjectNames();

        expect(tempProjectNames).toEqual([
          '@rush-temp/project1',
          '@rush-temp/project2',
          '@rush-temp/project3'
        ]);
      });
    }

    describe('V5.0 lockfile', () => {
      const filename: string = path.resolve(
        __dirname,
        '../../../src/logic/test/shrinkwrapFile/non-workspace-pnpm-lock-v5.yaml'
      );
      const shrinkwrapFile: BaseShrinkwrapFile = ShrinkwrapFileFactory.getShrinkwrapFile('pnpm', filename)!;

      validateNonWorkspaceLockfile(shrinkwrapFile);
    });

    describe('V5.3 lockfile', () => {
      const filename: string = path.resolve(
        __dirname,
        '../../../src/logic/test/shrinkwrapFile/non-workspace-pnpm-lock-v5.3.yaml'
      );
      const shrinkwrapFile: BaseShrinkwrapFile = ShrinkwrapFileFactory.getShrinkwrapFile('pnpm', filename)!;

      validateNonWorkspaceLockfile(shrinkwrapFile);
    });

    describe('V6.1 lockfile', () => {
      const filename: string = path.resolve(
        __dirname,
        '../../../src/logic/test/shrinkwrapFile/non-workspace-pnpm-lock-v6.1.yaml'
      );
      const shrinkwrapFile: BaseShrinkwrapFile = ShrinkwrapFileFactory.getShrinkwrapFile('pnpm', filename)!;

      validateNonWorkspaceLockfile(shrinkwrapFile);
    });

    describe('V9 lockfile', () => {
      const filename: string = path.resolve(
        __dirname,
        '../../../src/logic/test/shrinkwrapFile/non-workspace-pnpm-lock-v9.yaml'
      );
      const shrinkwrapFile: BaseShrinkwrapFile = ShrinkwrapFileFactory.getShrinkwrapFile('pnpm', filename)!;
      validateNonWorkspaceLockfile(shrinkwrapFile);
    });
  });

  describe('workspace', () => {
    let jsonSaveAsyncSpy: jest.SpyInstance;
    beforeEach(() => {
      jsonSaveAsyncSpy = jest.spyOn(JsonFile, 'saveAsync').mockReturnValue(Promise.resolve(true));
    });

    afterEach(() => {
      jsonSaveAsyncSpy.mockRestore();
    });

    function validateWorkspaceLockfile(shrinkwrapFile: BaseShrinkwrapFile): void {
      it('verifies project dependencies', async () => {
        const projectNames: string[] = ['project1', 'project2', 'project3'];
        for (const projectName of projectNames) {
          jsonSaveAsyncSpy.mockClear();
          const rushConfigurationProject: RushConfigurationProject = {
            projectRushTempFolder: `${projectName}/.rush/temp`,
            projectFolder: projectName,
            rushConfiguration: {
              commonTempFolder: 'common/temp'
            },
            subspace: {
              getSubspaceTempFolderPath: () => 'common/temp'
            }
          } as RushConfigurationProject;

          const projectShrinkwrap = shrinkwrapFile.getProjectShrinkwrap(rushConfigurationProject);
          await projectShrinkwrap?.updateProjectShrinkwrapAsync();
          expect(jsonSaveAsyncSpy).toHaveBeenCalledTimes(1);
          expect(jsonSaveAsyncSpy.mock.calls).toMatchSnapshot(projectName);
        }
      });

      it('does not have any temp projects', () => {
        const tempProjectNames: ReadonlyArray<string> = shrinkwrapFile.getTempProjectNames();
        expect(tempProjectNames).toHaveLength(0);
      });
    }

    describe('V5.3 lockfile', () => {
      const filename: string = path.resolve(
        __dirname,
        '../../../src/logic/test/shrinkwrapFile/workspace-pnpm-lock-v5.3.yaml'
      );
      const shrinkwrapFile: BaseShrinkwrapFile = ShrinkwrapFileFactory.getShrinkwrapFile('pnpm', filename)!;

      validateWorkspaceLockfile(shrinkwrapFile);
    });

    describe('V6.1 lockfile', () => {
      const filename: string = path.resolve(
        __dirname,
        '../../../src/logic/test/shrinkwrapFile/workspace-pnpm-lock-v5.3.yaml'
      );
      const shrinkwrapFile: BaseShrinkwrapFile = ShrinkwrapFileFactory.getShrinkwrapFile('pnpm', filename)!;

      validateWorkspaceLockfile(shrinkwrapFile);
    });

    describe('V9 lockfile', () => {
      const filename: string = path.resolve(
        __dirname,
        '../../../src/logic/test/shrinkwrapFile/workspace-pnpm-lock-v9.yaml'
      );

      const shrinkwrapFile: BaseShrinkwrapFile = ShrinkwrapFileFactory.getShrinkwrapFile('pnpm', filename)!;

      validateWorkspaceLockfile(shrinkwrapFile);
    });
  });
});

function testParsePnpmDependencyKey(packageName: string, key: string): string | undefined {
  const specifier: DependencySpecifier | undefined = parsePnpmDependencyKey(packageName, key);
  if (!specifier) {
    return undefined;
  }
  return specifier.versionSpecifier;
}

describe(parsePnpmDependencyKey.name, () => {
  it('extracts a simple version with no slashes', () => {
    expect(testParsePnpmDependencyKey('anonymous', '0.0.5')).toEqual('0.0.5');
  });
  it('extracts a simple package name', () => {
    expect(testParsePnpmDependencyKey('isarray', '/isarray/2.0.5')).toEqual('2.0.5');
    expect(testParsePnpmDependencyKey('@scope/test-dep', '/@scope/test-dep/1.2.3-beta.3')).toEqual(
      '1.2.3-beta.3'
    );
  });
  it('extracts a registry-qualified path', () => {
    expect(
      testParsePnpmDependencyKey('@scope/test-dep', 'example.pkgs.visualstudio.com/@scope/test-dep/1.0.0')
    ).toEqual('1.0.0');
    expect(
      testParsePnpmDependencyKey(
        '@scope/test-dep',
        'example.pkgs.visualstudio.com/@scope/test-dep/1.2.3-beta.3'
      )
    ).toEqual('1.2.3-beta.3');
  });
  it('extracts a V3 peer dependency path', () => {
    expect(testParsePnpmDependencyKey('gulp-karma', '/gulp-karma/0.0.5/karma@0.13.22')).toEqual('0.0.5');
    expect(testParsePnpmDependencyKey('sinon-chai', '/sinon-chai/2.8.0/chai@3.5.0+sinon@1.17.7')).toEqual(
      '2.8.0'
    );
    expect(
      testParsePnpmDependencyKey('@ms/sp-client-utilities', '/@ms/sp-client-utilities/3.1.1/foo@13.1.0')
    ).toEqual('3.1.1');
    expect(
      testParsePnpmDependencyKey(
        'tslint-microsoft-contrib',
        '/tslint-microsoft-contrib/6.2.0/tslint@5.18.0+typescript@3.5.3'
      )
    ).toEqual('6.2.0');
  });
  it('extracts a V5 peer dependency path', () => {
    expect(testParsePnpmDependencyKey('anonymous', '23.6.0_babel-core@6.26.3')).toEqual('23.6.0');
    expect(testParsePnpmDependencyKey('anonymous', '1.0.7_request@2.88.0')).toEqual('1.0.7');
    expect(testParsePnpmDependencyKey('anonymous', '1.0.3_@pnpm+logger@1.0.2')).toEqual('1.0.3');
    expect(
      testParsePnpmDependencyKey(
        'tslint-microsoft-contrib',
        '/tslint-microsoft-contrib/6.2.0_tslint@5.18.0+typescript@3.5.3'
      )
    ).toEqual('6.2.0');
  });
  it('detects NPM package aliases', () => {
    expect(testParsePnpmDependencyKey('alias1', '/isarray/2.0.5')).toEqual('npm:isarray@2.0.5');
    expect(testParsePnpmDependencyKey('alias2', '/@ms/sp-client-utilities/3.1.1/foo@13.1.0')).toEqual(
      'npm:@ms/sp-client-utilities@3.1.1'
    );
  });
  it('detects urls', () => {
    expect(
      testParsePnpmDependencyKey('example', '@github.com/abc/def/abcdef2fbd0260e6e56ed5ba34df0f5b6599bbe64')
    ).toEqual('@github.com/abc/def/abcdef2fbd0260e6e56ed5ba34df0f5b6599bbe64');
    expect(
      testParsePnpmDependencyKey('example', 'github.com/abc/def/abcdef2fbd0260e6e56ed5ba34df0f5b6599bbe64')
    ).toEqual('github.com/abc/def/abcdef2fbd0260e6e56ed5ba34df0f5b6599bbe64');
    expect(
      testParsePnpmDependencyKey('example', 'bitbucket.com/abc/def/abcdef2fbd0260e6e56ed5ba34df0f5b6599bbe64')
    ).toEqual('bitbucket.com/abc/def/abcdef2fbd0260e6e56ed5ba34df0f5b6599bbe64');
    expect(
      testParsePnpmDependencyKey(
        'example',
        'microsoft.github.com/abc/def/abcdef2fbd0260e6e56ed5ba34df0f5b6599bbe64'
      )
    ).toEqual('microsoft.github.com/abc/def/abcdef2fbd0260e6e56ed5ba34df0f5b6599bbe64');
    expect(
      testParsePnpmDependencyKey(
        'example',
        'microsoft.github/.com/abc/def/abcdef2fbd0260e6e56ed5ba34df0f5b6599bbe64'
      )
    ).toEqual('microsoft.github/.com/abc/def/abcdef2fbd0260e6e56ed5ba34df0f5b6599bbe64');
    expect(
      testParsePnpmDependencyKey('example', 'ab.cd.ef.gh/ijkl/abcdef2fbd0260e6e56ed5ba34df0f5b6599bbe64')
    ).toEqual('ab.cd.ef.gh/ijkl/abcdef2fbd0260e6e56ed5ba34df0f5b6599bbe64');
    expect(testParsePnpmDependencyKey('example', 'ab.cd/ef')).toEqual('ab.cd/ef');
  });
  it('handles bad cases', () => {
    expect(testParsePnpmDependencyKey('example', '/foo/gulp-karma/0.0.5/karma@0.13.22')).toEqual(undefined);
    expect(testParsePnpmDependencyKey('example', '/@ms/3.1.1/foo@13.1.0')).toEqual(undefined);
    expect(testParsePnpmDependencyKey('example', 'file:projects/my-app.tgz')).toEqual(undefined);
    expect(testParsePnpmDependencyKey('example', '')).toEqual(undefined);
    expect(testParsePnpmDependencyKey('example', '/')).toEqual(undefined);
    expect(testParsePnpmDependencyKey('example', '//')).toEqual(undefined);
    expect(testParsePnpmDependencyKey('example', '/@/')).toEqual(undefined);
    expect(testParsePnpmDependencyKey('example', 'example.pkgs.visualstudio.com/@scope/testDep/')).toEqual(
      undefined
    );
    expect(
      testParsePnpmDependencyKey(
        'example',
        'microsoft.github.com/abc\\def/abcdef2fbd0260e6e56ed5ba34df0f5b6599bbe64'
      )
    ).toEqual(undefined);
    expect(
      testParsePnpmDependencyKey(
        'example',
        'microsoft.github.com/abc/def//abcdef2fbd0260e6e56ed5ba34df0f5b6599bbe64'
      )
    ).toEqual(undefined);
    expect(
      testParsePnpmDependencyKey(
        'example',
        'microsoft./github.com/abc/def/abcdef2fbd0260e6e56ed5ba34df0f5b6599bbe64'
      )
    ).toEqual(undefined);
    expect(
      testParsePnpmDependencyKey(
        'example',
        'microsoft/abc/github/abc/def/abcdef2fbd0260e6e56ed5ba34df0f5b6599bbe64'
      )
    ).toEqual(undefined);
    expect(testParsePnpmDependencyKey('example', 'ab.cd/ef/')).toEqual(undefined);
  });
});
