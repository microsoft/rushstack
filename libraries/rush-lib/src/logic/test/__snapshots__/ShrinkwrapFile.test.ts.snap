// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PnpmShrinkwrapFile workspace V5.3 lockfile verifies project dependencies: project1 1`] = `
Array [
  Array [
    Object {
      "../../project1": "../../project1:D5ar2j+w6/zH15/eOoF37Nkdbamt2tX47iijyj7LVXk=:",
      "/jquery/1.12.3": "sha512-FzM42/Ew+Hb8ha2OlhHRBLgWIZS32gZ0+NvWTf+ZvVvGaIlJkOiXQyb7VBjv4L6fJfmTrRf3EsAmbfsHDhfemw==",
      "/pad-left/1.0.2": "sha512-saxSV1EYAytuZDtQYEwi0DPzooG6aN18xyHrnJtzwjVwmMauzkEecd7hynVJGolNGk1Pl9tltmZqfze4TZTCxg==",
      "/repeat-string/1.6.1": "sha512-PV0dzCYDNfRi1jCDbJzpW7jNNDRuCOG/jI5ctQcGKt/clZD+YcPS3yIlWuTJMmESC8aevCFmWJy5wjAFgNqN6w==",
    },
    "project1/.rush/temp/shrinkwrap-deps.json",
    Object {
      "ensureFolderExists": true,
    },
  ],
]
`;

exports[`PnpmShrinkwrapFile workspace V5.3 lockfile verifies project dependencies: project2 1`] = `
Array [
  Array [
    Object {
      "../../project2": "../../project2:PQ2FvyHHwmt/FIUaiJBVAfpHv6hj9EGJrAw69+0IY50=:",
      "/jquery/2.2.4": "sha512-lBHj60ezci2u1v2FqnZIraShGgEXq35qCzMv4lITyHGppTnA13rwR0MgwyNJh9TnDs3aXUvd1xjAotfraMHX/Q==",
      "/q/1.5.1": "sha512-kV/CThkXo6xyFEZUugw/+pIOywXcDbFYgSct5cT3gqlbkBE1SJdwy6UQoZvodiWF/ckQLZyDE/Bu1M6gVu5lVw==",
    },
    "project2/.rush/temp/shrinkwrap-deps.json",
    Object {
      "ensureFolderExists": true,
    },
  ],
]
`;

exports[`PnpmShrinkwrapFile workspace V5.3 lockfile verifies project dependencies: project3 1`] = `
Array [
  Array [
    Object {
      "../../project3": "../../project3:jomsZKvXG32qqYOfW3HUBGfWWSw6ybFV1WDf9c/kiP4=:",
      "/q/1.5.1": "sha512-kV/CThkXo6xyFEZUugw/+pIOywXcDbFYgSct5cT3gqlbkBE1SJdwy6UQoZvodiWF/ckQLZyDE/Bu1M6gVu5lVw==",
      "/repeat-string/1.6.1": "sha512-PV0dzCYDNfRi1jCDbJzpW7jNNDRuCOG/jI5ctQcGKt/clZD+YcPS3yIlWuTJMmESC8aevCFmWJy5wjAFgNqN6w==",
      "example.pkgs.visualstudio.com/@scope/testDep/2.1.0": "example.pkgs.visualstudio.com/@scope/testDep/2.1.0:i5jbUOp/IxUgiN0dHYsTRzmimOVBwu7f9908rRaC9VY=:",
    },
    "project3/.rush/temp/shrinkwrap-deps.json",
    Object {
      "ensureFolderExists": true,
    },
  ],
]
`;

exports[`PnpmShrinkwrapFile workspace V6.1 lockfile verifies project dependencies: project1 1`] = `
Array [
  Array [
    Object {
      "../../project1": "../../project1:D5ar2j+w6/zH15/eOoF37Nkdbamt2tX47iijyj7LVXk=:",
      "/jquery/1.12.3": "sha512-FzM42/Ew+Hb8ha2OlhHRBLgWIZS32gZ0+NvWTf+ZvVvGaIlJkOiXQyb7VBjv4L6fJfmTrRf3EsAmbfsHDhfemw==",
      "/pad-left/1.0.2": "sha512-saxSV1EYAytuZDtQYEwi0DPzooG6aN18xyHrnJtzwjVwmMauzkEecd7hynVJGolNGk1Pl9tltmZqfze4TZTCxg==",
      "/repeat-string/1.6.1": "sha512-PV0dzCYDNfRi1jCDbJzpW7jNNDRuCOG/jI5ctQcGKt/clZD+YcPS3yIlWuTJMmESC8aevCFmWJy5wjAFgNqN6w==",
    },
    "project1/.rush/temp/shrinkwrap-deps.json",
    Object {
      "ensureFolderExists": true,
    },
  ],
]
`;

exports[`PnpmShrinkwrapFile workspace V6.1 lockfile verifies project dependencies: project2 1`] = `
Array [
  Array [
    Object {
      "../../project2": "../../project2:PQ2FvyHHwmt/FIUaiJBVAfpHv6hj9EGJrAw69+0IY50=:",
      "/jquery/2.2.4": "sha512-lBHj60ezci2u1v2FqnZIraShGgEXq35qCzMv4lITyHGppTnA13rwR0MgwyNJh9TnDs3aXUvd1xjAotfraMHX/Q==",
      "/q/1.5.1": "sha512-kV/CThkXo6xyFEZUugw/+pIOywXcDbFYgSct5cT3gqlbkBE1SJdwy6UQoZvodiWF/ckQLZyDE/Bu1M6gVu5lVw==",
    },
    "project2/.rush/temp/shrinkwrap-deps.json",
    Object {
      "ensureFolderExists": true,
    },
  ],
]
`;

exports[`PnpmShrinkwrapFile workspace V6.1 lockfile verifies project dependencies: project3 1`] = `
Array [
  Array [
    Object {
      "../../project3": "../../project3:jomsZKvXG32qqYOfW3HUBGfWWSw6ybFV1WDf9c/kiP4=:",
      "/q/1.5.1": "sha512-kV/CThkXo6xyFEZUugw/+pIOywXcDbFYgSct5cT3gqlbkBE1SJdwy6UQoZvodiWF/ckQLZyDE/Bu1M6gVu5lVw==",
      "/repeat-string/1.6.1": "sha512-PV0dzCYDNfRi1jCDbJzpW7jNNDRuCOG/jI5ctQcGKt/clZD+YcPS3yIlWuTJMmESC8aevCFmWJy5wjAFgNqN6w==",
      "example.pkgs.visualstudio.com/@scope/testDep/2.1.0": "example.pkgs.visualstudio.com/@scope/testDep/2.1.0:i5jbUOp/IxUgiN0dHYsTRzmimOVBwu7f9908rRaC9VY=:",
    },
    "project3/.rush/temp/shrinkwrap-deps.json",
    Object {
      "ensureFolderExists": true,
    },
  ],
]
`;

exports[`PnpmShrinkwrapFile workspace V9 lockfile verifies project dependencies: project1 1`] = `
Array [
  Array [
    Object {
      "../../project1": "../../project1:6yFTI2g+Ny0Au80xpo6zIY61TCNDUuLUd6EgLlbOBtc=:",
      "jquery@1.12.3": "sha512-FzM42/Ew+Hb8ha2OlhHRBLgWIZS32gZ0+NvWTf+ZvVvGaIlJkOiXQyb7VBjv4L6fJfmTrRf3EsAmbfsHDhfemw==",
      "pad-left@1.0.2": "sha512-saxSV1EYAytuZDtQYEwi0DPzooG6aN18xyHrnJtzwjVwmMauzkEecd7hynVJGolNGk1Pl9tltmZqfze4TZTCxg==",
      "repeat-string@1.6.1": "sha512-PV0dzCYDNfRi1jCDbJzpW7jNNDRuCOG/jI5ctQcGKt/clZD+YcPS3yIlWuTJMmESC8aevCFmWJy5wjAFgNqN6w==",
    },
    "project1/.rush/temp/shrinkwrap-deps.json",
    Object {
      "ensureFolderExists": true,
    },
  ],
]
`;

exports[`PnpmShrinkwrapFile workspace V9 lockfile verifies project dependencies: project2 1`] = `
Array [
  Array [
    Object {
      "../../project2": "../../project2:l6v/HWUhScMI0m4k6D5qHiCOFj3Z0GoIFJEcp4I63w0=:",
      "jquery@2.2.4": "sha512-lBHj60ezci2u1v2FqnZIraShGgEXq35qCzMv4lITyHGppTnA13rwR0MgwyNJh9TnDs3aXUvd1xjAotfraMHX/Q==",
      "q@1.5.0": "sha512-VVMcd+HnuWZalHPycK7CsbVJ+sSrrrnCvHcW38YJVK9Tywnb5DUWJjONi81bLUj7aqDjIXnePxBl5t1r/F/ncg==",
    },
    "project2/.rush/temp/shrinkwrap-deps.json",
    Object {
      "ensureFolderExists": true,
    },
  ],
]
`;

exports[`PnpmShrinkwrapFile workspace V9 lockfile verifies project dependencies: project3 1`] = `
Array [
  Array [
    Object {
      "../../project3": "../../project3:vMoje8cXfsHYOc6EXbxEw/qyBGXGUL1RApmNfwl7oA8=:",
      "pad-left@https://github.com/jonschlinkert/pad-left/tarball/2.1.0": "pad-left@https://github.com/jonschlinkert/pad-left/tarball/2.1.0:bKrL+SvVYubL0HwTq/GOOXq1d05LTQ+HGqlXabzGEAU=:",
      "q@1.5.0": "sha512-VVMcd+HnuWZalHPycK7CsbVJ+sSrrrnCvHcW38YJVK9Tywnb5DUWJjONi81bLUj7aqDjIXnePxBl5t1r/F/ncg==",
      "repeat-string@1.6.1": "sha512-PV0dzCYDNfRi1jCDbJzpW7jNNDRuCOG/jI5ctQcGKt/clZD+YcPS3yIlWuTJMmESC8aevCFmWJy5wjAFgNqN6w==",
    },
    "project3/.rush/temp/shrinkwrap-deps.json",
    Object {
      "ensureFolderExists": true,
    },
  ],
]
`;
